File: MoviesMVVMAPPWithPagingJava/app/src/main/java/com/masterclass/moviesmvvmappwithpagingjava/viewModels/MoviesViewModel.java
package com.masterclass.moviesmvvmappwithpagingjava.viewModels;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelKt;
import androidx.paging.Pager;
import androidx.paging.PagingConfig;
import androidx.paging.PagingData;
import androidx.paging.rxjava3.PagingRx;

import com.masterclass.moviesmvvmappwithpagingjava.models.Movie;
import com.masterclass.moviesmvvmappwithpagingjava.paging.MoviePagingSource;

import io.reactivex.rxjava3.core.Flowable;
import kotlinx.coroutines.CoroutineScope;

public class MoviesViewModel extends ViewModel {
    Flowable<PagingData<Movie>> flowableMoviePagingData;

    public MoviesViewModel(){
        init();
    }

    public void init(){

        //Define Paging Source
        MoviePagingSource moviePagingSource = new MoviePagingSource();

        Pager<Integer, Movie> pager = new Pager(
            new PagingConfig(
                    20,
                    20,
                    false,
                    20,
                    20*499
            ), () -> moviePagingSource);

        //Flowable
        flowableMoviePagingData = PagingRx.getFlowable(pager);
        CoroutineScope coroutineScope = ViewModelKt.getViewModelScope(this);
        PagingRx.cachedIn(flowableMoviePagingData, coroutineScope);
    }
}

File: MoviesMVVMAPPWithPagingKotlin/app/src/main/java/com/masterclass/moviesmvvmappwithpagingkotlin/viewModels/MoviesViewModel.kt
package com.masterclass.moviesmvvmappwithpagingkotlin.viewModels

import androidx.lifecycle.ViewModel
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import com.masterclass.moviesmvvmappwithpagingkotlin.models.Movie
import com.masterclass.moviesmvvmappwithpagingkotlin.paging.MoviePagingSource
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

class MoviesViewModel : ViewModel() {
    var flowableMoviePagingData: Flow<PagingData<Movie>>? = null
    init {
        init()
    }

    private fun init() {

        //Define Paging Source
        val moviePagingSource = MoviePagingSource()
        val pager = Pager(
            PagingConfig(
                pageSize = 20,
                prefetchDistance = 20,
                enablePlaceholders = false,
                initialLoadSize = 20,
                maxSize = 20 * 499
            )
        ) { moviePagingSource }

        //Flowable
        flowableMoviePagingData = pager.flow
            .map { pagingData -> pagingData }
    }
}
